conftest.py

```diff
@@ -25,4 +25,7 @@ def pytest_runtest_setup(item):
     else:  # Skip GPU tests if cupy not installed
         from importlib.util import find_spec
 
-        pytest.mark.skipif(not find_spec("cupy"), reason="No cupy installed")
+        if not find_spec("cupy"):
+            pytest.skip("Cupy not installed.")
+        # why is this different?
+        # pytest.mark.skipif(not find_spec("cupy"), reason="No cupy installed")
```

Comment:
Because marks can be either used as decorators or by passing them to the `marks=` kwarg of `pytest.param`.

They can’t be used imperatively. You can do `item.add_marker(pytest.mark.skipif(...))`, but as said, instead of fiddling with this, you can just do this for the tests

```py
@needs('cupy')
@pytest.mark.gpu
def test_...(): ...
```

and this in pyproject.toml to deselect them by default:

```toml
[tool.pytest.ini_options]
addopts = [
    ...
    '-m', 'not gpu',
]
```

all tests can be run with `pytest -m 'gpu and not gpu'`

Reply:
I don't think we should deselect these by default. They should be tested if the current environment has the ability to test them. We may be able to define this just a marker then, but will look into it later.